default/python/AI/ProductionAI.py

```diff
@@ -1402,10 +1402,37 @@ def generate_production_orders():
                 if (not prioritized) and (this_priority == PriorityType.PRODUCTION_INVASION):
                     fo.issueRequeueProductionOrder(production_queue.size - 1, 0)  # move to front
         print
+    update_stockpile_use()
     fo.updateProductionQueue()
     _print_production_queue(after_turn=True)
 
 
+def update_stockpile_use():
+    """Decide which elements in the production_queue will be enabled for drawing from the imperial stockpile.  This
+    initial version simply ensures that every resource group with at least one item on the queue has its highest
+    priority item be stockpile-enabled.
+
+    :return: None
+    """
```

Comment:
The current implementation allows to have a low-priority item in the queue that is stockpile-enabled:

Consider there is a short disruption in the supply lines. The disconnected system enqueues some filler item to get rid of excess PP in that system but that item is not really needed. Because it is disconnected, it will draw from stockpile.
After supply connection is restored, the item is still enabled for stockpile-usage. 

So, we may want to add a loop prior to this implementation that stockpile-disables all the queue items.

Reply:
If we were going to do that I think that a simpler way to get the same treatment would be to handle it within the same loop, checking and turning off the stockpile as part of the ```if element.locationID in planets_with_group_stockpile_elem:``` check

But, I had considered that route and chose not to take it, I'd be more open to simply not enabling the stockpile for items outside the group that includes the imperial capital.  Especially as we expand this code, but even now, I think it's fine and good that on turn N the top priority item would get marked as stockpile enabled, and then if on turn N+1 another item is made top priority and the former top slips to #2 priority, it is still stockpile enabled.

I think that the real problem in your scenario is that an unwanted filler item is being built at all-- under both your proposal and the current PR the "filler" item will still be getting stockpile enabled while the side system is disconnected, and also under both approaches it will continue getting built and drawing PP once the system reconnects to the main group, the only difference is the _potential_ for some extra stockpile drain, and I don't see avoiding that being worth the cost of restricting the main group to having just its top item be stockpile enabled.

Let's look a little more closely at the "filler" item.  My recollection is that such can only be ships.  Outpost ships still be valuable no matter where they are built (at least currently), and to a fair extent I would expect the same to be true for colony ships (which the AI would be very unlikely to build anyways).   In an isolated system either of those two ship types might not be the most advanced available to the empire, but should still be fine to build.  The real potential problem right now is with military ships and to a lesser extent troop ships.  

Right now there is a check for military ships, so that if we start building one with relatively poor pilots it is logged, but it still gets built.  I think that's the bit of could that should be augmented for this problem-- no need to directly look at pilot, but for military ships and troop ships if the rating/cost efficiency number there is too low then we just don't build such things there. That threshold could be at 50% or 70%, somewhere in there, barring exceptional circumstances.